Pods/MapCache/MapCache/Classes/CachedTileOverlay.swift

```diff
@@ -10,43 +10,105 @@ import MapKit
+            // print("loadTile:: not using cache")
             return super.loadTile(at: path, result: result)
         } else {
-           return mapCache.loadTile(at: path, result: result)
+            return mapCache.loadTile(at: path, result: result)
+        }
+    }
+    
+    ///
+    /// Tells whether or not to upsample and show a lesser detailed z level
+    /// takes into account `useZoom` configuration as well as current and `maximumZ` values
+    /// - Parameter at current zoom.
+    func shouldZoom(at scale: MKZoomScale) -> Bool {
+        guard mapCache.config.overZoomMaximumZ else { return false }
+        return scale.toZoomLevel(tileSize: mapCache.config.tileSize) > mapCache.config.maximumZ
+    }
+    
+    ///
+    /// Given the maximum zoom level for the tileset `(mapCache.config.maximumZ`) it will return the tile, map rects, and additional scaling factor for upscaling tiles.
+    ///
+    /// - Parameter rect: map rectangle for which we want to get the tile set
+    /// - Parameter scale: current zoom scale
+    ///
+    func tilesInMapRect(rect: MKMapRect, scale: MKZoomScale) -> [ZoomableTile] {
+        var tiles: [ZoomableTile] = []
+        let tileSize = mapCache.config.tileSize
+        var z = scale.toZoomLevel(tileSize: tileSize)
+       
+        // Represents the number of tiles the current tile is going to be divided
+        var overZoom = 1
+        let tileSetMaxZ = mapCache.config.maximumZ
+        if (z > tileSetMaxZ) {
+            overZoom = Int(pow(2.0, Double(z - tileSetMaxZ)))
+            z = tileSetMaxZ
+        }
+        
+        let adjustedTileSize = Double(overZoom * Int(tileSize.width))
+        
+        let minX = Int(floor((rect.minX * Double(scale)) / adjustedTileSize))
+        let maxX = Int(floor((rect.maxX * Double(scale)) / adjustedTileSize))
+        let minY = Int(floor((rect.minY * Double(scale)) / adjustedTileSize))
+        let maxY = Int(floor((rect.maxY * Double(scale)) / adjustedTileSize))
+        
+        for x in minX ... maxX {
+            for y in minY ... maxY {
+                
+                let point = MKMapPoint(x: (Double(x) * adjustedTileSize) / Double(scale),
+                                       y: (Double(y) * adjustedTileSize) / Double(scale))
+                let size = MKMapSize(width: adjustedTileSize / Double(scale),
+                                     height: adjustedTileSize / Double(scale))
+                let tileRect = MKMapRect(origin: point, size: size)
+                
+                // check that a portion of the tile intersects with the maps rect
+                // no need to do the work on a tile that won't be seen
+                guard rect.intersects(tileRect) else { continue }
+                
+                let path =  MKTileOverlayPath(x: x, y: y, z: z, contentScaleFactor: scale)
+                let tile = ZoomableTile(maximumZPath: path, rect: tileRect, overZoom: Zoom(overZoom))
```

Comment:
Got a crash at

Reply:
Only seems to happen if the tile I'm using is OpenTopoMap, zooming in as much as I could. Tested and replicable on both iOS 13.4.1 and iOS 14 GM

Reply:
I believe it happens as `overzoom` can reach 256 / 512, which is more than what `UInt8` `Zoom` can handle.